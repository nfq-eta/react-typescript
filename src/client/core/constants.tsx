export const ACTION = {
    CHECKBOX: {
        ADD: 'ac@checkbox.add',
    },
};
