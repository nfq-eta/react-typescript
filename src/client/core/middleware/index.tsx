import logger from 'redux-logger';
export {
    logger,
};
