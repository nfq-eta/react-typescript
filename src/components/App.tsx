import * as React from 'react';
import CheckBox, { IItem } from './CheckBox';

import CheckboxItemsJson = require('./tests/__mocks__/CheckboxItems.json');
import CheckboxSelectedItemsJson = require('./tests/__mocks__/CheckboxSelectedItems.json');

export interface IAppProps {
    title?: string;
}

export interface IAppState {
    items: IItem[];
}

export default class App extends React.Component<IAppProps, IAppState> {
    selectedItems = new Map(CheckboxSelectedItemsJson);

    constructor() {
        super();

        this.state = { items: CheckboxItemsJson };
    }

    handleClick = (item: IItem, checked: boolean) => {
        if (checked) {
            this.selectedItems.set(item.id, item);
        } else {
            this.selectedItems.delete(item.id);
        }
    };

    isSelected(item: IItem): boolean {
        return this.selectedItems.has(item.id);
    }

    render() {
        return (
            <div>
                <h1>{this.props.title || ''}</h1>
                {this.state.items.map(item => (
                    <CheckBox
                        item={item}
                        key={item.id}
                        checked={this.isSelected(item) || false}
                        handleClick={this.handleClick}
                    />
                    ))
                }
            </div>
        );
    }
}
