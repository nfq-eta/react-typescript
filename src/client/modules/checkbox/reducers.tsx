import { Action, handleActions } from 'redux-actions';
import { IItem } from '../../components/checkbox/CheckBox';
import { ACTION } from '../../core/constants';

const initialState: IItem[] = [];

export default handleActions(
    {
        [ACTION.CHECKBOX.ADD]: (state: IItem[], action: Action<IItem>) => {
            return [...state, action.payload];
        },
    },

    initialState,
);
