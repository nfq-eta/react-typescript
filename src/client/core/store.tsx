import { createStore, applyMiddleware, Store } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import { logger } from './middleware';

import rootReducer, { IRootState } from './reducers';
import { IItem } from '../components/CheckBox';

export function configureStore(initialState?: IRootState) {
    let middleware = applyMiddleware(logger);

    if (process.env.NODE_ENV === 'development') {
        middleware = composeWithDevTools(middleware);
    }

    const store = createStore(rootReducer, initialState, middleware) as Store<IItem[]>;

    if (module.hot) {
        module.hot.accept('./reducers', () => {
            const nextReducer = require('./reducers');
            store.replaceReducer(nextReducer);
        });
    }

    return store;
}
