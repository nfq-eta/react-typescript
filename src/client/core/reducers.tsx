import { combineReducers } from 'redux';
import checkBoxReducers from '../modules/checkbox/reducers';
import { IItem } from '../components/checkbox/CheckBox';

export interface IRootState {
    checkBoxItems: IItem[];
}

const rootReducer = combineReducers({
    checkBoxItems: checkBoxReducers,
});

export default rootReducer;
