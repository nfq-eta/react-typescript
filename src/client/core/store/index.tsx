import { store as createStore } from 'react-easy-state';
import { IItem } from '../../components/checkbox/CheckBoxComponent';

export type rootStore = {
  foo: string;
  checkBoxItems: IItem[];
};

export const rootStore: rootStore = {
  foo: 'buzz',
  checkBoxItems: [],
};

export const store = createStore({
  ...rootStore,
  // tslint:disable-next-line
  ...typeof window !== 'undefined' && window && window.__PRELOADED_STATE__,
  addAction(item: IItem) {
    store.checkBoxItems.push(item);
  },
});
