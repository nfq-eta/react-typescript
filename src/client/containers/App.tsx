import * as React from 'react';
import { bindActionCreators } from 'redux';
import { connect, MapDispatchToProps } from 'react-redux';
import * as uniqId from 'uniqid';

import CheckBox, { IItem } from '../components/checkbox/CheckBox';
import * as CheckboxItemsJson from '../components/checkbox/tests/__mocks__/CheckboxItems.json';
import * as CheckboxSelectedItemsJson from '../components/checkbox/tests/__mocks__/CheckboxSelectedItems.json';
import { addCheckbox } from '../modules/checkbox/actions';
import { IRootState } from '../core/reducers';

export interface IAppProps {
    addAction: typeof addCheckbox;
    items: IItem[];
}

export interface IAppState {
    items: IItem[];
}

// @connect(mapStateToProps, mapDispatchToProps)
class App extends React.Component<IAppProps, IAppState> {
    selectedItems = new Map(CheckboxSelectedItemsJson);

    constructor() {
        super();

        this.state = { items: CheckboxItemsJson };
    }

    handleClick = (item: IItem, checked: boolean) => {
        if (checked) {
            this.selectedItems.set(item.id, item);
        } else {
            this.selectedItems.delete(item.id);
        }
    };

    isSelected(item: IItem): boolean {
        return this.selectedItems.has(item.id);
    }

    handleAdd = () => {
        this.props.addAction({
            id: uniqId(),
            value: 'demo',
            label: 'Demo',
        });
    };

    render() {
        return (
            <div>
                <h1>App</h1>
                <button onClick={this.handleAdd}>Add more</button>
                {this.props.items.map(item => (
                    <div key={item.id}>
                        <CheckBox
                            item={item}
                            checked={this.isSelected(item) || false}
                            handleClick={this.handleClick}
                        />
                    </div>
                ))}
            </div>
        );
    }
}

export function mapStateToProps(state: IRootState) {
    return ({
        items: state.checkBoxItems,
    });
}

export function mapDispatchToProps(dispatch: MapDispatchToProps<any, any>) {
    return ({
        addAction: bindActionCreators(addCheckbox, dispatch),
    });
}

export default connect(mapStateToProps, mapDispatchToProps)(App);

export { App as AppDisconnected };
