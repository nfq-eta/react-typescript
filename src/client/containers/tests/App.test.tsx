import * as React from 'react';
import { shallow } from 'enzyme';
import enzymeToJson from 'enzyme-to-json';
import {} from 'jest';

import { AppDisconnected, mapStateToProps } from '../App';
import { IRootState } from '../../core/reducers';

describe('App.tsx', () => {
    const props = {
        items: [
            { id: 'demo1', label: 'Demo', value: 'demo' },
        ],
        addAction: jest.fn(),
    };

    it('renders without crashing', () => {
        shallow(<AppDisconnected {...props} />);
    });

    it('matches snapshot', () => {
        const wrapper = shallow(<AppDisconnected {...props} />);
        expect(enzymeToJson(wrapper)).toMatchSnapshot();
    });

    it('Title should be visible', () => {
        const wrapper = shallow(<AppDisconnected {...props} />);
        expect(wrapper.find('h1').text()).toContain('App');
    });

    it('should handle click', () => {
        const wrapper = shallow(<AppDisconnected {...props} />);
        const instance = wrapper.instance() as AppDisconnected;
        const params = { id: 'test', value: 1, label: 'Demo' };

        instance.handleClick(params, true);
        expect(instance.selectedItems.has('test')).toBeTruthy();

        instance.handleClick(params, false);
        expect(instance.selectedItems.has('test')).toBeFalsy();
    });

    it('handleAdd()', () => {
        const wrapper = shallow(<AppDisconnected {...props} />);
        const instance = wrapper.instance() as AppDisconnected;

        instance.handleAdd();

        expect(props.addAction).toBeCalledWith({ id: expect.any(String), label: 'Demo', value: 'demo' });
    });

    it('mapStateToProps()', () => {
        const initialState: IRootState = {
            checkBoxItems: [],
        };

        expect(mapStateToProps(initialState)).toEqual({ items: [] });
    });

});
