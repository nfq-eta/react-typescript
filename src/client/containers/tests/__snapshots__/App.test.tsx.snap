// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.tsx matches snapshot 1`] = `
<div>
  <h1>
    App
  </h1>
  <button
    onClick={[Function]}
  >
    Add more
  </button>
  <div
    key="demo1"
  >
    <CheckBox
      checked={false}
      handleClick={[Function]}
      item={
        Object {
          "id": "demo1",
          "label": "Demo",
          "value": "demo",
        }
      }
    />
  </div>
</div>
`;
